React Intermediate Challenge 1 - Styles, Fetching API, Testing

soyN3sh opened the pull request:
- Adds title, date picker and picture section
- Adds NASA API fetch call
- Adds tests

![Screen Shot 2021-11-01 at 15 49 03](https://user-images.githubusercontent.com/92320673/139746893-cee64d1a-60c0-437d-a174-4b9067860f05.png)

soyNesh committed:
React Intermediate Challenge 1 - Styles, Fetching API, Testing

- Adds title, date picker and picture section
- Adds NASA API fetch call
- Adds tests

AlanMalagon commented on challenge-1/picture-of-the-day/package.json:
```diff
@@ -6,9 +6,11 @@
     "@testing-library/jest-dom": "^5.14.1",
     "@testing-library/react": "^11.2.7",
     "@testing-library/user-event": "^12.8.3",
+    "date-fns": "^2.25.0",
```
Veo que tomaste mi consejo de usar esta librería, bien allí ahorra mucho tiempo y es confiable.

soyN3sh replied:
Está chidori. La otra que conocía es moment.js

AlanMalagon commented on challenge-1/picture-of-the-day/src/__tests__/App.test.jsx:
```diff
@@ -0,0 +1,101 @@
+    });
+  });
+
+  it("renders correctly when API call error", (done) => {
+    const tmpState = {
+      state: {
+        apiParams: {
+          date: "2021-10-30",
+        },
+        error: {
+          exists: true,
+        },
+      },
+    };
+
+    mockFetchError();
+    act(async () => {
+      render(<AppWithContext state={tmpState} dispatch={mockDispatch} />);
+    }).then(() => {
+      const errorMessage = document.querySelector("#errorMessage");
+      expect(errorMessage).toBeInTheDocument();
+      done();
+    });
+  });
```
Veo que agregaste también este tests después de la sesión que tuvimos, está perfecto, personalmente yo también habría agregado el expect de que se muestre el error que regresa el mockFetchError, para asegurarnos que lo imprime en pantalla.

soyN3sh replied:
Listo, agregado.

AlanMalagon commented on challenge-1/picture-of-the-day/src/components/DatePicker/DatePicker.component.jsx:
```diff
@@ -0,0 +1,33 @@
+  const handleDateChange = (e) => {
+    dispatch({
+      type: actions.setApiParams,
+      payload: { date: format(parseISO(e.target.value), "yyyy-MM-dd") },
```
No tengo completo contexto de esto, pero me parece a mi que te habrías ahorrado esto de haber establecido inicialmente el state del input como string en tu state inicial. Los input de type date pueden recibir ya strings y sino me equivoco el formato por default es el que justo nos piden usar.

soyN3sh replied:
Lo dejé así "pa no errarle"

AlanMalagon commented on challenge-1/picture-of-the-day/src/components/DatePicker/DatePicker.component.jsx:
```diff
@@ -0,0 +1,33 @@
+
+  return (
+    <MyDatePicker>
+      <DpInput
```
Creo que personalmente habría usado el nombre `Input` tal cual, a fin de cuentas estás definiendo el type después y el nombre del componente ya indica la principal función fuera del componente, mientras que aquí en el componente en si es bueno tener el concepto concreto de que este es un Input básico y maleable.

soyN3sh replied:
Listo, le cambié el nombre. Sí se ve mejor.

AlanMalagon commented on challenge-1/picture-of-the-day/src/components/DatePicker/DatePicker.styles.jsx:
```diff
@@ -0,0 +1,43 @@
+  float: left;
+  position: relative;
+
+  * {
+    -webkit-touch-callout: none; /* iOS Safari */
+    -webkit-user-select: none; /* Chrome/Safari/Opera */
+    -khtml-user-select: none; /* Konqueror */
+    -moz-user-select: none; /* Firefox */
+    -ms-user-select: none; /* Internet Explorer/Edge */
+    user-select: none; /* Non-prefixed version, currently */
+  }
```
No sé bien qué hace esta propiedad de CSS pero el hecho de usar styled-components ya agrega las diferentes propiedades para compatibilidad con navegadores. Por favor confírmame si estoy en lo correcto cuando puedas porque no conozco esta propiedad y quizás se nueva, de ser así se justificaría que mantengamos estas lineas.

soyN3sh replied:
Es para que no se pueda seleccionar el texto. Fue parte de unos estilos que vi por ahí.

AlanMalagon commented on challenge-1/picture-of-the-day/src/components/DatePicker/DatePicker.styles.jsx:
```diff
@@ -0,0 +1,43 @@
+  }
+`;
+
+const DpInput = styled.input.attrs((props) => ({
+  type: "date",
+}))`
```
Creo que esta propiedad podría dejarse sin ningún problema del lado del component en lugar de styles, pero hey, no sabía que se podía hacer esto, así que aprendemos los dos cositas nuevas.

soyN3sh replied:
Sí está chidori también. En realidad nada más lo dejé de este lado porque también lo acababa de descubrir y me gustó que se veía más limpio del otro lado.

AlanMalagon commented on challenge-1/picture-of-the-day/src/components/DatePicker/DatePicker.styles.jsx:
```diff
@@ -0,0 +1,43 @@
+  overflow: hidden;
+  border-radius: 20px;
+
+  input:focus {
+    outline: none;
+  }
+
+  input {
+    width: 125%;
+    background: #f5f5f5;
+    border: none;
+    height: 35px;
+    text-align: center;
+    text-transform: uppercase;
+    letter-spacing: 2px;
+    font-size: 11px;
+    cursor: pointer;
+  }
```
Quizás no estoy viendo algo aquí, pero sí ya establecimos en la variable que esto va a ser un `styled.input` porque estamos haciendo que seleccione todos los inputs dentro de este contenedor con estas líneas?

soyN3sh replied:
Listo, ya lo dejé directo 👍

AlanMalagon commented on challenge-1/picture-of-the-day/src/components/Picture/Picture.component.jsx:
```diff
@@ -0,0 +1,36 @@
+import GlobalContext from "../../providers/GlobalContext";
+import useData from "../../utils/hooks/useData";
+
+const Picture = () => {
+  const {
+    state: { apiParams, error },
+  } = useContext(GlobalContext);
+
+  const { data } = useData(apiParams);
+
+  if (error.exists) {
+    return <CustomError id="errorMessage">Error: {error.message}</CustomError>;
+  }
+
+  return data ? (
+    <PictureContainer>
+      <CustomImg id="customImg" src={data.url} alt={data.title} />
+      <CustomSubtitle>{data.title}</CustomSubtitle>
+      <CustomAuthor>By: {data.copyright}</CustomAuthor>
+      <CustomParagraph>{data.explanation}</CustomParagraph>
+    </PictureContainer>
+  ) : (
+    "loading..."
+  );
+};
```
Que bueno que pudimos discutir sobre esto en la reunión y refactorizarlo. Quedó mucho más limpio, perfecto para darle mantenimiento en el futuro sin perdernos entre if statements.

soyN3sh replied:
Pero muuuucho más limpio jejetl también ya le agregué un loader más cool

AlanMalagon commented on challenge-1/picture-of-the-day/src/providers/GlobalContext.jsx:
```diff
@@ -0,0 +1,5 @@
+import { createContext } from 'react';
+
+const GlobalContext = createContext(null);
+
+export default GlobalContext;
```
Que bueno que aplicaste el context en la aplicación, yo no lo vi necesario ayer que estaba intentando hacer el reto pero honestamente nos ahorró mucho tiempo cuando hicimos la refactorización del `Picture.component`

soyN3sh replied:
Sí llegué a pensar que tal vez no era necesario, pero preferí aplicarlo para manejar más fácil el state.

AlanMalagon reviewed (approved):
Muy buen trabajo amigo, esta vez te noté más suelto con los conceptos de React, me agrada que nuestro enfoque ahora haya sido más en cuanto a buenas prácticas y testing que React en si, solo confirma lo que ya te decía, que las bases ya las tienes firmes para resolver problemas utilizando React como herramienta.

soyNesh committed:
Adds FoldingCube loader to Picture component

soyNesh committed:
Attends PR comments

- Adds expect looking for error message
- Renames DbInput to just Input for better reading
- Removes unnecessary input selector from DatePicker.styles
